File: java/src/com/haibo/yan/algorithm/RotateArray.java
package com.haibo.yan.algorithm;

import java.util.Arrays;

import static java.util.stream.Collectors.joining;

public class RotateArray {
    public  void rotateRight(int[] array, int k) {
        if (array.length == 0) return;

        int rk = k % array.length;

        reverse(array, 0, array.length - 1 - rk);
        reverse(array, array.length - rk, array.length - 1);
        reverse(array, 0, array.length - 1);
    }

    private void reverse(int[] array, int start, int end) {
        if (start + 1 >= end) {
            return;
        }

        int middle = (start + end) / 2;

        for (int i = start; i <= middle; i++) {
            int t = array[i];
            array[i] = array[end + start - i];
            array[end + start - i] = t;
        }
    }

    public static void main(String[] args) {
        RotateArray instance = new RotateArray();
        int[] testArray1 = new int[]{1, 2, 3};

        instance.rotateRight(testArray1, 4);

        System.out.println(Arrays.stream(testArray1).mapToObj(String::valueOf).collect(joining(", ", "[", "]")));
    }
}

File: scala/src/main/scala/com/haiboyan/scala/algorithm/array/RotateArray.scala
package com.haiboyan.scala.algorithm.array

object RotateArray extends App{
  def rotateLeft(array : Array[Int], k: Int) : Array[Int] = {
    if (array.length == 0) {
      array
    } else {
      val m = k % array.length
      (array.slice(0, m).reverse ++ array.slice(m, array.length).reverse).reverse
    }
  }

  def rotateRight(array : Array[Int], k: Int) = rotateLeft(array, array.length - k)
}
